File: vibemediaplayer/src/auraphyx/renderer.ts
import { BLOOM_FRAG_SRC, SCENE_FRAG_SRC, VERT_SRC } from './shaders';
import type { ShaderUniforms } from './types';

// Two triangles covering the full [-1,1] clip-space quad.
const QUAD = new Float32Array([
  -1, -1, 1, -1, 1, 1,
  -1, -1, 1, 1, -1, 1,
]);

function compileShader(gl: WebGL2RenderingContext, type: number, source: string): WebGLShader {
  const shader = gl.createShader(type);
  if (!shader) throw new Error('failed to create shader');
  gl.shaderSource(shader, source);
  gl.compileShader(shader);
  if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
    const log = gl.getShaderInfoLog(shader);
    gl.deleteShader(shader);
    throw new Error(`shader compile failed: ${log}`);
  }
  return shader;
}

function linkProgram(gl: WebGL2RenderingContext, vertSrc: string, fragSrc: string): WebGLProgram {
  const vert = compileShader(gl, gl.VERTEX_SHADER, vertSrc);
  const frag = compileShader(gl, gl.FRAGMENT_SHADER, fragSrc);
  const program = gl.createProgram();
  if (!program) throw new Error('failed to create program');
  gl.attachShader(program, vert);
  gl.attachShader(program, frag);
  gl.linkProgram(program);
  gl.deleteShader(vert);
  gl.deleteShader(frag);
  if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
    const log = gl.getProgramInfoLog(program);
    gl.deleteProgram(program);
    throw new Error(`program link failed: ${log}`);
  }
  return program;
}

type SceneUniformLocs = {
  time: WebGLUniformLocation | null;
  resolution: WebGLUniformLocation | null;
  centroid: WebGLUniformLocation | null;
  saturation: WebGLUniformLocation | null;
  syncopation: WebGLUniformLocation | null;
  bpmNorm: WebGLUniformLocation | null;
  groove: WebGLUniformLocation | null;
  dissonance: WebGLUniformLocation | null;
  valence: WebGLUniformLocation | null;
  arousal: WebGLUniformLocation | null;
  scenarioFog: WebGLUniformLocation | null;
  primaryRgb: WebGLUniformLocation | null;
  secondaryRgb: WebGLUniformLocation | null;
  brightnessFloor: WebGLUniformLocation | null;
  brightnessCeiling: WebGLUniformLocation | null;
  strobeTrigger: WebGLUniformLocation | null;
  fogDensity: WebGLUniformLocation | null;
  visualNoise: WebGLUniformLocation | null;
  chromEnergy: WebGLUniformLocation | null;
  entrainment: WebGLUniformLocation | null;
  tslX: WebGLUniformLocation | null;
  tslY: WebGLUniformLocation | null;
  tslZ: WebGLUniformLocation | null;
  phaseAlign: WebGLUniformLocation | null;
  latticeRot: WebGLUniformLocation | null;
  auraphyxMode: WebGLUniformLocation | null;
};

type BloomUniformLocs = {
  scene: WebGLUniformLocation | null;
  resolution: WebGLUniformLocation | null;
  bloomStrength: WebGLUniformLocation | null;
  fadeAmount: WebGLUniformLocation | null;
};

/** Two-pass Chladni field (+ Auraphyx lattice) -> bloom composite renderer. */
export class AuraphyxRenderer {
  private gl: WebGL2RenderingContext;
  private sceneProgram: WebGLProgram;
  private bloomProgram: WebGLProgram;
  private sceneU: SceneUniformLocs;
  private bloomU: BloomUniformLocs;
  private quadVao: WebGLVertexArrayObject;
  private fbo: WebGLFramebuffer;
  private sceneTexture: WebGLTexture;
  private width = 1;
  private height = 1;

  bloomEnabled = true;
  visualNoise = false;
  auraphyxEnabled = true;

  constructor(gl: WebGL2RenderingContext, width: number, height: number) {
    this.gl = gl;
    this.sceneProgram = linkProgram(gl, VERT_SRC, SCENE_FRAG_SRC);
    this.bloomProgram = linkProgram(gl, VERT_SRC, BLOOM_FRAG_SRC);
    this.sceneU = this.locateSceneUniforms();
    this.bloomU = this.locateBloomUniforms();
    this.quadVao = this.buildQuad();
    this.sceneTexture = gl.createTexture()!;
    this.fbo = gl.createFramebuffer()!;
    this.resize(width, height);
  }

  private locateSceneUniforms(): SceneUniformLocs {
    const gl = this.gl;
    const p = this.sceneProgram;
    const at = (name: string) => gl.getUniformLocation(p, name);
    return {
      time: at('u_time'),
      resolution: at('u_resolution'),
      centroid: at('u_centroid'),
      saturation: at('u_saturation'),
      syncopation: at('u_syncopation'),
      bpmNorm: at('u_bpm_norm'),
      groove: at('u_groove'),
      dissonance: at('u_dissonance'),
      valence: at('u_valence'),
      arousal: at('u_arousal'),
      scenarioFog: at('u_scenario_fog'),
      primaryRgb: at('u_primary_rgb'),
      secondaryRgb: at('u_secondary_rgb'),
      brightnessFloor: at('u_brightness_floor'),
      brightnessCeiling: at('u_brightness_ceiling'),
      strobeTrigger: at('u_strobe_trigger'),
      fogDensity: at('u_fog_density'),
      visualNoise: at('u_visual_noise'),
      chromEnergy: at('u_chrom_energy[0]'),
      entrainment: at('u_entrainment'),
      tslX: at('u_tsl_x'),
      tslY: at('u_tsl_y'),
      tslZ: at('u_tsl_z'),
      phaseAlign: at('u_phase_align'),
      latticeRot: at('u_lattice_rot'),
      auraphyxMode: at('u_auraphyx_mode'),
    };
  }

  private locateBloomUniforms(): BloomUniformLocs {
    const gl = this.gl;
    const p = this.bloomProgram;
    const at = (name: string) => gl.getUniformLocation(p, name);
    return {
      scene: at('u_scene'),
      resolution: at('u_resolution'),
      bloomStrength: at('u_bloom_strength'),
      fadeAmount: at('u_fade_amount'),
    };
  }

  private buildQuad(): WebGLVertexArrayObject {
    const gl = this.gl;
    const vao = gl.createVertexArray();
    if (!vao) throw new Error('failed to create VAO');
    gl.bindVertexArray(vao);
    const vbo = gl.createBuffer();
    gl.bindBuffer(gl.ARRAY_BUFFER, vbo);
    gl.bufferData(gl.ARRAY_BUFFER, QUAD, gl.STATIC_DRAW);
    gl.enableVertexAttribArray(0);
    gl.vertexAttribPointer(0, 2, gl.FLOAT, false, 0, 0);
    gl.bindVertexArray(null);
    return vao;
  }

  resize(width: number, height: number): void {
    const gl = this.gl;
    this.width = Math.max(1, width | 0);
    this.height = Math.max(1, height | 0);

    gl.bindTexture(gl.TEXTURE_2D, this.sceneTexture);
    gl.texImage2D(
      gl.TEXTURE_2D, 0, gl.RGBA8, this.width, this.height, 0,
      gl.RGBA, gl.UNSIGNED_BYTE, null,
    );
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);

    gl.bindFramebuffer(gl.FRAMEBUFFER, this.fbo);
    gl.framebufferTexture2D(
      gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.sceneTexture, 0,
    );
    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
    gl.bindTexture(gl.TEXTURE_2D, null);
  }

  /** `time` is a locally-owned clock (seconds), independent of the backend's
   * own `uniforms.time` — smoother than pacing motion to the ~60Hz emit rate. */
  render(uniforms: ShaderUniforms, time: number): void {
    const gl = this.gl;

    // ---- Pass 1: Chladni + Auraphyx field -> FBO texture ----
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.fbo);
    gl.viewport(0, 0, this.width, this.height);
    gl.clearColor(0, 0, 0, 1);
    gl.clear(gl.COLOR_BUFFER_BIT);

    gl.useProgram(this.sceneProgram);
    const u = this.sceneU;
    gl.uniform1f(u.time, time);
    gl.uniform2f(u.resolution, this.width, this.height);
    gl.uniform1f(u.centroid, uniforms.centroid);
    gl.uniform1f(u.saturation, uniforms.saturation);
    gl.uniform1f(u.syncopation, uniforms.syncopation);
    gl.uniform1f(u.bpmNorm, uniforms.bpm_norm);
    gl.uniform1f(u.groove, uniforms.groove);
    gl.uniform1f(u.dissonance, uniforms.dissonance);
    gl.uniform1f(u.valence, uniforms.valence);
    gl.uniform1f(u.arousal, uniforms.arousal);
    gl.uniform1f(u.scenarioFog, 0.0);
    gl.uniform3fv(u.primaryRgb, uniforms.primary_rgb);
    gl.uniform3fv(u.secondaryRgb, uniforms.secondary_rgb);
    gl.uniform1f(u.brightnessFloor, uniforms.brightness_floor);
    gl.uniform1f(u.brightnessCeiling, uniforms.brightness_ceiling);
    gl.uniform1f(u.strobeTrigger, uniforms.strobe_trigger);
    gl.uniform1f(u.fogDensity, uniforms.fog_density);
    gl.uniform1f(u.visualNoise, this.visualNoise ? 1.0 : 0.0);
    gl.uniform1fv(u.chromEnergy, uniforms.chrom_energy);
    gl.uniform1f(u.entrainment, uniforms.entrainment);
    gl.uniform1f(u.tslX, uniforms.tsl_x);
    gl.uniform1f(u.tslY, uniforms.tsl_y);
    gl.uniform1f(u.tslZ, uniforms.tsl_z);
    gl.uniform1f(u.phaseAlign, uniforms.phase_align);
    gl.uniform1f(u.latticeRot, uniforms.lattice_rot);
    gl.uniform1f(u.auraphyxMode, this.auraphyxEnabled ? 1.0 : 0.0);

    gl.bindVertexArray(this.quadVao);
    gl.drawArrays(gl.TRIANGLES, 0, 6);

    // ---- Pass 2: bloom composite -> screen (default framebuffer) ----
    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
    gl.viewport(0, 0, this.width, this.height);
    gl.clearColor(0, 0, 0, 1);
    gl.clear(gl.COLOR_BUFFER_BIT);

    gl.useProgram(this.bloomProgram);
    const b = this.bloomU;
    gl.activeTexture(gl.TEXTURE0);
    gl.bindTexture(gl.TEXTURE_2D, this.sceneTexture);
    gl.uniform1i(b.scene, 0);
    gl.uniform2f(b.resolution, this.width, this.height);
    gl.uniform1f(b.bloomStrength, this.bloomEnabled ? bloomStrength(uniforms) : 0.0);
    gl.uniform1f(b.fadeAmount, 1.0);

    gl.bindVertexArray(this.quadVao);
    gl.drawArrays(gl.TRIANGLES, 0, 6);
    gl.bindVertexArray(null);
  }

  dispose(): void {
    const gl = this.gl;
    gl.deleteFramebuffer(this.fbo);
    gl.deleteTexture(this.sceneTexture);
    gl.deleteVertexArray(this.quadVao);
    gl.deleteProgram(this.sceneProgram);
    gl.deleteProgram(this.bloomProgram);
  }
}

/** Bloom strength: arousal x brightness_ceiling, matching `VapRuntime::bloom_strength()`. */
function bloomStrength(uniforms: ShaderUniforms): number {
  return uniforms.arousal * uniforms.brightness_ceiling * 1.5;
}

File: vap-claude/vibe-visualizer/src/renderer.rs
//! Phase 4 — GL renderer, using `glium`. Two-pass pipeline mirroring
//! `src/gl_renderer.c`: Pass 1 renders the Chladni field (vibe.frag) to an
//! off-screen FBO texture; Pass 2 runs the bloom post-process
//! (post_bloom.frag) reading that texture and composing to the screen.

use glium::backend::glutin::Display;
use glium::glutin::surface::WindowSurface;
use glium::index::{NoIndices, PrimitiveType};
use glium::texture::Texture2d;
use glium::uniforms::{MagnifySamplerFilter, MinifySamplerFilter};
use glium::{implement_vertex, uniform, Program, Surface, VertexBuffer};

use vap_core::{NexusVibezFrame, VapRuntime};

const VERT_SRC: &str = include_str!("../shaders/vibe.vert");
const SCENE_FRAG_SRC: &str = include_str!("../shaders/vibe.frag");
const BLOOM_FRAG_SRC: &str = include_str!("../shaders/post_bloom.frag");

#[derive(Copy, Clone)]
struct Vertex {
    a_pos: [f32; 2],
}
implement_vertex!(Vertex, a_pos);

// Two triangles covering the full [-1,1] clip-space quad.
const QUAD: [Vertex; 6] = [
    Vertex { a_pos: [-1.0, -1.0] },
    Vertex { a_pos: [1.0, -1.0] },
    Vertex { a_pos: [1.0, 1.0] },
    Vertex { a_pos: [-1.0, -1.0] },
    Vertex { a_pos: [1.0, 1.0] },
    Vertex { a_pos: [-1.0, 1.0] },
];

pub struct RenderState {
    width: u32,
    height: u32,
    pub bloom_enabled: bool,
    pub visual_noise: bool,
    scene_program: Program,
    bloom_program: Program,
    quad_vbo: VertexBuffer<Vertex>,
    scene_texture: Texture2d,
    time_accum: f32,
}

impl RenderState {
    pub fn new(display: &Display<WindowSurface>, width: u32, height: u32) -> anyhow::Result<Self> {
        let scene_program = Program::from_source(display, VERT_SRC, SCENE_FRAG_SRC, None)
            .map_err(|e| anyhow::anyhow!("failed to compile scene shader: {e}"))?;
        let bloom_program = Program::from_source(display, VERT_SRC, BLOOM_FRAG_SRC, None)
            .map_err(|e| anyhow::anyhow!("failed to compile bloom shader: {e}"))?;
        let quad_vbo = VertexBuffer::new(display, &QUAD)?;
        let scene_texture = Texture2d::empty(display, width.max(1), height.max(1))?;

        Ok(RenderState {
            width: width.max(1),
            height: height.max(1),
            bloom_enabled: true,
            visual_noise: false,
            scene_program,
            bloom_program,
            quad_vbo,
            scene_texture,
            time_accum: 0.0,
        })
    }

    pub fn resize(&mut self, display: &Display<WindowSurface>, width: u32, height: u32) -> anyhow::Result<()> {
        self.width = width.max(1);
        self.height = height.max(1);
        self.scene_texture = Texture2d::empty(display, self.width, self.height)?;
        Ok(())
    }

    /// Render one frame: Pass 1 (Chladni + Nexus Vibez field) to the FBO,
    /// Pass 2 (bloom composite) to the screen.
    pub fn render(
        &mut self,
        display: &Display<WindowSurface>,
        vap: &VapRuntime,
        nexus: &NexusVibezFrame,
        dt: f32,
    ) -> anyhow::Result<()> {
        self.time_accum += dt;
        let resolution = [self.width as f32, self.height as f32];
        let indices = NoIndices(PrimitiveType::TrianglesList);

        // ── Pass 1: Cymatic field render -> FBO color texture ──────────
        {
            let mut fbo = glium::framebuffer::SimpleFrameBuffer::new(display, &self.scene_texture)?;
            fbo.clear_color(0.0, 0.0, 0.0, 1.0);

            let chrom = vap.chroma_energy;
            let uniforms = uniform! {
                u_time: self.time_accum,
                u_resolution: resolution,
                u_centroid: vap.spectral_centroid_hz,
                u_saturation: vap.saturation_index,
                u_syncopation: vap.syncopation_index,
                u_bpm_norm: vap.bpm_norm(),
                u_groove: vap.groove_quantization,
                u_dissonance: vap.dissonance_density,
                u_valence: vap.affective.valence,
                u_arousal: vap.affective.arousal,
                u_scenario_fog: vap.scenario_fog(),
                u_primary_rgb: vap.photometric.primary_hex,
                u_secondary_rgb: vap.photometric.secondary_hex,
                u_brightness_floor: vap.photometric.brightness_floor,
                u_brightness_ceiling: vap.photometric.brightness_ceiling,
                u_strobe_trigger: vap.photometric.strobe_threshold,
                u_fog_density: vap.photometric.fog_density,
                u_visual_noise: if self.visual_noise { 1.0f32 } else { 0.0f32 },
                u_entrainment: vap.entrainment_factor,
                u_tsl_x: nexus.tsl_x,
                u_tsl_y: nexus.tsl_y,
                u_tsl_z: nexus.tsl_z,
                u_phase_align: nexus.phase_align,
                u_lattice_rot: nexus.lattice_rot,
            }
            // `uniform float u_chrom_energy[4]` is a GLSL array, which OpenGL
            // addresses per-element by name ("name[i]") — glium's fixed-size
            // [f32; N] AsUniformValue impls map to vecN instead, so each
            // element is bound individually here to match the array uniform.
            .add("u_chrom_energy[0]", chrom[0])
            .add("u_chrom_energy[1]", chrom[1])
            .add("u_chrom_energy[2]", chrom[2])
            .add("u_chrom_energy[3]", chrom[3]);

            fbo.draw(
                &self.quad_vbo,
                indices,
                &self.scene_program,
                &uniforms,
                &Default::default(),
            )?;
        }

        // ── Pass 2: Bloom post-process -> screen ────────────────────────
        let mut target = display.draw();
        target.clear_color(0.0, 0.0, 0.0, 1.0);

        let sampler = self
            .scene_texture
            .sampled()
            .magnify_filter(MagnifySamplerFilter::Linear)
            .minify_filter(MinifySamplerFilter::Linear);
        let bloom_strength = if self.bloom_enabled { vap.bloom_strength() } else { 0.0 };
        let bloom_uniforms = uniform! {
            u_scene: sampler,
            u_resolution: resolution,
            u_bloom_strength: bloom_strength,
            u_fade_amount: 1.0f32,
        };
        target.draw(
            &self.quad_vbo,
            indices,
            &self.bloom_program,
            &bloom_uniforms,
            &Default::default(),
        )?;
        target.finish()?;

        Ok(())
    }
}
